pqarrow: panic on serializing arrow to parquet

I see the following panic after we enabled arrow ingestion:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x60 pc=0x17d289f]

goroutine 10483 [running]:
github.com/apache/arrow/go/v10/arrow/array.(*Dictionary).Dictionary(...)
	github.com/apache/arrow/go/v10@v10.0.1/arrow/array/dictionary.go:232
github.com/polarsignals/frostdb/pqarrow.appendToRow({0xc0e42f0000, 0x1, 0x29}, {0x2c133a0, 0xc1527539a0?}, 0xc, 0xc000d80c00?, 0xc1086f1b80?, 0x0?)
	github.com/polarsignals/frostdb@v0.0.0-20230424150726-5288c29ad8a0/pqarrow/parquet.go:59 +0xcff
github.com/polarsignals/frostdb/pqarrow.getRecordRow(0xd?, 0xd?, {0x2c105c8, 0xc120178d20}, 0xa?, {0xc1086f1b80, 0x29, 0x3?}, {0xc037651000, 0x29, ...})
	github.com/polarsignals/frostdb@v0.0.0-20230424150726-5288c29ad8a0/pqarrow/parquet.go:90 +0x15a
github.com/polarsignals/frostdb/pqarrow.RecordToFile(0xc19b2ea700, 0xc14224baa0, {0x2c105c8, 0xc120178d20})
	github.com/polarsignals/frostdb@v0.0.0-20230424150726-5288c29ad8a0/pqarrow/parquet.go:179 +0x2ef
github.com/polarsignals/frostdb/parts.(*Part).SerializeBuffer(0xc19b2ea700?, 0x2bd9fc0?, 0xc1512679b0?)
	github.com/polarsignals/frostdb@v0.0.0-20230424150726-5288c29ad8a0/parts/part.go:53 +0x33
github.com/polarsignals/frostdb/parts.(*Part).AsSerializedBuffer(0xc152769040, 0xc19b2ea700)
	github.com/polarsignals/frostdb@v0.0.0-20230424150726-5288c29ad8a0/parts/part.go:69 +0x10a
github.com/polarsignals/frostdb.compactLevel0IntoLevel1(0xc03916efc0, 0xe35?, {0xc148950000, 0xe35, 0x1400}, {0x0, 0x0, 0x0?}, 0x3ff0000000000000, 0xc0989c9b60)
	github.com/polarsignals/frostdb@v0.0.0-20230424150726-5288c29ad8a0/compaction.go:559 +0x3de
github.com/polarsignals/frostdb.(*TableBlock).Serialize.func2({0x2bdf1a0?, 0xc00a4955f0?})
	github.com/polarsignals/frostdb@v0.0.0-20230424150726-5288c29ad8a0/table.go:1512 +0x1ac
github.com/google/btree.(*node[...]).iterate(0x2c331e0, 0xc145a98b40, {{0x0?, 0x0?}, 0x7?}, {{0x0?, 0x0?}, 0x1?}, 0x0, 0x0, ...)
	github.com/google/btree@v1.1.2/btree_generic.go:522 +0x3f7
github.com/google/btree.(*BTreeG[...]).Ascend(0xc0dee99030?, 0xc19b2d58c0?)
	github.com/google/btree@v1.1.2/btree_generic.go:779 +0x49
github.com/google/btree.(*BTree).Ascend(...)
	github.com/google/btree@v1.1.2/btree_generic.go:1029
github.com/polarsignals/frostdb.(*TableBlock).Serialize(0xc03916efc0, {0x2be3860, 0xc183c701d0})
	github.com/polarsignals/frostdb@v0.0.0-20230424150726-5288c29ad8a0/table.go:1500 +0x487
github.com/polarsignals/frostdb.(*TableBlock).Persist.func1()
	github.com/polarsignals/frostdb@v0.0.0-20230424150726-5288c29ad8a0/store.go:40 +0x6c
created by github.com/polarsignals/frostdb.(*TableBlock).Persist
	github.com/polarsignals/frostdb@v0.0.0-20230424150726-5288c29ad8a0/store.go:38 +0x2a5
```

Which looks like it panics on this line:
https://github.com/apache/arrow/blob/9009dd76e4e9a2f4f13340ebf4173e71813b359b/go/arrow/array/dictionary.go#L234

It looks like maybe somehow the dictionary is not initialized properly?

```
panic: runtime error: index out of range [0] with length 0

goroutine 34012 [running]:
github.com/apache/arrow/go/v10/arrow/array.(*simpleRecord).Column(0xc00f8b05b0?, 0xc16aa87998?)
	github.com/apache/arrow/go/v10@v10.0.1/arrow/array/record.go:206 +0x3a
github.com/polarsignals/frostdb/pqarrow.getRecordRow(0xd?, 0xd?, {0x2c10868, 0xc1a46ef200}, 0xa?, {0xc016213a00, 0x10, 0x3?}, {0xc16aef0580, 0x10, ...})
	github.com/polarsignals/frostdb@v0.0.0-20230425155736-e1792af839f6/pqarrow/parquet.go:90 +0x124
github.com/polarsignals/frostdb/pqarrow.RecordToFile(0xc00f8b05b0, 0xc174f78a80, {0x2c10868, 0xc1a46ef200})
	github.com/polarsignals/frostdb@v0.0.0-20230425155736-e1792af839f6/pqarrow/parquet.go:179 +0x2ef
github.com/polarsignals/frostdb/parts.(*Part).SerializeBuffer(0xc00f8b05b0?, 0x2bda260?, 0xc01378f6e0?)
	github.com/polarsignals/frostdb@v0.0.0-20230425155736-e1792af839f6/parts/part.go:53 +0x33
github.com/polarsignals/frostdb/parts.(*Part).AsSerializedBuffer(0xc00996c190, 0xc00f8b05b0)
	github.com/polarsignals/frostdb@v0.0.0-20230425155736-e1792af839f6/parts/part.go:69 +0x10a
github.com/polarsignals/frostdb.(*TableBlock).Serialize.func1.1(0x40?)
	github.com/polarsignals/frostdb@v0.0.0-20230425155736-e1792af839f6/table.go:1468 +0x48
github.com/polarsignals/frostdb.(*Granule).PartsForTx.func1(0xc1f60f3558?)
	github.com/polarsignals/frostdb@v0.0.0-20230425155736-e1792af839f6/granule.go:102 +0x3d
github.com/polarsignals/frostdb/parts.(*List).Iterate(0xc16983b0a0, 0xc0b3b2f9b0)
	github.com/polarsignals/frostdb@v0.0.0-20230425155736-e1792af839f6/parts/part_list.go:90 +0x5f
github.com/polarsignals/frostdb.(*Granule).PartsForTx(0x0?, 0x3000100?, 0x0?)
	github.com/polarsignals/frostdb@v0.0.0-20230425155736-e1792af839f6/granule.go:96 +0x47
github.com/polarsignals/frostdb.(*TableBlock).Serialize.func1({0x2bdf440?, 0xc1bf965200?})
	github.com/polarsignals/frostdb@v0.0.0-20230425155736-e1792af839f6/table.go:1467 +0x85
github.com/google/btree.(*node[...]).iterate(0x2c33480, 0xc00a182ad0, {{0x0?, 0x0?}, 0xff?}, {{0x0?, 0x0?}, 0x5?}, 0x0, 0x1, ...)
	github.com/google/btree@v1.1.2/btree_generic.go:522 +0x3f7
github.com/google/btree.(*node[...]).iterate(0x2c33480, 0xc000c85db0, {{0x0?, 0x0?}, 0xff?}, {{0x0?, 0x0?}, 0x5?}, 0x0, 0x1, ...)
	github.com/google/btree@v1.1.2/btree_generic.go:510 +0x23f
github.com/google/btree.(*node[...]).iterate(0x2c33480, 0xc013a73580, {{0x0?, 0x0?}, 0xff?}, {{0x0?, 0x0?}, 0x5?}, 0x0, 0x1, ...)
	github.com/google/btree@v1.1.2/btree_generic.go:510 +0x23f
github.com/google/btree.(*node[...]).iterate(0x2c33480, 0x0, {{0x0?, 0x0?}, 0x0?}, {{0x0?, 0x0?}, 0x5d?}, 0x0, 0x1, ...)
	github.com/google/btree@v1.1.2/btree_generic.go:510 +0x23f
github.com/google/btree.(*node[...]).iterate(0x2c33480, 0xc1f1decb20, {{0x0?, 0x0?}, 0x7?}, {{0x0?, 0x0?}, 0x80?}, 0x0, 0x0, ...)
	github.com/google/btree@v1.1.2/btree_generic.go:527 +0x4be
github.com/google/btree.(*BTreeG[...]).Ascend(0x416010?, 0xbe4?)
	github.com/google/btree@v1.1.2/btree_generic.go:779 +0x49
github.com/google/btree.(*BTree).Ascend(...)
	github.com/google/btree@v1.1.2/btree_generic.go:1029
github.com/polarsignals/frostdb.(*TableBlock).Serialize(0xc007362900, {0x2be3b00, 0xc03da1ba28})
	github.com/polarsignals/frostdb@v0.0.0-20230425155736-e1792af839f6/table.go:1464 +0x15e
github.com/polarsignals/frostdb.(*TableBlock).Persist.func1()
	github.com/polarsignals/frostdb@v0.0.0-20230425155736-e1792af839f6/store.go:40 +0x6c
created by github.com/polarsignals/frostdb.(*TableBlock).Persist
	github.com/polarsignals/frostdb@v0.0.0-20230425155736-e1792af839f6/store.go:38 +0x2a5
	```
	
Copying from issue #427 as I'm merging that issue into this one due to the same root cause.

--- a/snapshot.go
+++ b/snapshot.go
@@ -306,83 +306,91 @@ func writeSnapshot(ctx context.Context, tx uint64, db *DB, w io.Writer) error {
 
 	metadata := &snapshotpb.FooterData{}
 	for _, t := range tables {
-		block := t.ActiveBlock()
-		blockUlid, err := block.ulid.MarshalBinary()
-		if err != nil {
-			return err
-		}
-		tableMeta := &snapshotpb.Table{
-			Name:   t.name,
-			Config: t.config,
-			ActiveBlock: &snapshotpb.Table_TableBlock{
-				Ulid:   blockUlid,
-				Size:   block.Size(),
-				MinTx:  block.minTx,
-				PrevTx: block.prevTx,
-			},
-		}
+		if err := func() error {
+			// Obtain a write block to prevent racing with
+			// compaction/persistence.
+			block, done, err := t.ActiveWriteBlock()
+			if err != nil {
+				return err
+			}
+			defer done()
+			blockUlid, err := block.ulid.MarshalBinary()
+			if err != nil {
+				return err
+			}
+			tableMeta := &snapshotpb.Table{
+				Name:   t.name,
+				Config: t.config,
+				ActiveBlock: &snapshotpb.Table_TableBlock{
+					Ulid:   blockUlid,
+					Size:   block.Size(),
+					MinTx:  block.minTx,
+					PrevTx: block.prevTx,
+				},
+			}
 
-		var ascendErr error
-		t.ActiveBlock().Index().Ascend(func(i btree.Item) bool {
-			granuleMeta := &snapshotpb.Granule{}
-			i.(*Granule).PartsForTx(tx, func(p *parts.Part) bool {
-				partMeta := &snapshotpb.Part{
-					StartOffset:     int64(offW.offset),
-					Tx:              p.TX(),
-					CompactionLevel: uint64(p.CompactionLevel()),
+			var ascendErr error
+			block.Index().Ascend(func(i btree.Item) bool {
+				granuleMeta := &snapshotpb.Granule{}
+				i.(*Granule).PartsForTx(tx, func(p *parts.Part) bool {
+					partMeta := &snapshotpb.Part{
+						StartOffset:     int64(offW.offset),
+						Tx:              p.TX(),
+						CompactionLevel: uint64(p.CompactionLevel()),
+					}
+					if err := func() error {
+						if err := ctx.Err(); err != nil {
+							return err
+						}
+						schema := t.schema
+
+						if record := p.Record(); record != nil {
+							partMeta.Encoding = snapshotpb.Part_ENCODING_ARROW
+							recordWriter := ipc.NewWriter(
+								w,
+								ipc.WithSchema(record.Schema()),
+							)
+							defer recordWriter.Close()
+							return recordWriter.Write(record)
+						}
+						partMeta.Encoding = snapshotpb.Part_ENCODING_PARQUET
+
+						buf, err := p.AsSerializedBuffer(schema)
+						if err != nil {
+							return err
+						}
+						rows := buf.Reader()
+						defer rows.Close()
+
+						parquetWriter, err := schema.GetWriter(w, buf.DynamicColumns())
+						if err != nil {
+							return err
+						}
+						defer schema.PutWriter(parquetWriter)
+						defer parquetWriter.Close()
+
+						if _, err := parquet.CopyRows(parquetWriter, rows); err != nil {
+							return err
+						}
+						return nil
+					}(); err != nil {
+						ascendErr = err
+						return false
+					}
+					partMeta.EndOffset = int64(offW.offset)
+					granuleMeta.PartMetadata = append(granuleMeta.PartMetadata, partMeta)
+					return true
+				})
+				if len(granuleMeta.PartMetadata) > 0 {
+					tableMeta.GranuleMetadata = append(tableMeta.GranuleMetadata, granuleMeta)
 				}
-				if err := func() error {
-					if err := ctx.Err(); err != nil {
-						return err
-					}
-					schema := t.schema
-
-					if record := p.Record(); record != nil {
-						partMeta.Encoding = snapshotpb.Part_ENCODING_ARROW
-						recordWriter := ipc.NewWriter(
-							w,
-							ipc.WithSchema(record.Schema()),
-						)
-						defer recordWriter.Close()
-						return recordWriter.Write(record)
-					}
-					partMeta.Encoding = snapshotpb.Part_ENCODING_PARQUET
-
-					buf, err := p.AsSerializedBuffer(schema)
-					if err != nil {
-						return err
-					}
-					rows := buf.Reader()
-					defer rows.Close()
-
-					parquetWriter, err := schema.GetWriter(w, buf.DynamicColumns())
-					if err != nil {
-						return err
-					}
-					defer schema.PutWriter(parquetWriter)
-					defer parquetWriter.Close()
-
-					if _, err := parquet.CopyRows(parquetWriter, rows); err != nil {
-						return err
-					}
-					return nil
-				}(); err != nil {
-					ascendErr = err
-					return false
-				}
-				partMeta.EndOffset = int64(offW.offset)
-				granuleMeta.PartMetadata = append(granuleMeta.PartMetadata, partMeta)
 				return true
 			})
-			if len(granuleMeta.PartMetadata) > 0 {
-				tableMeta.GranuleMetadata = append(tableMeta.GranuleMetadata, granuleMeta)
-			}
-			return true
-		})
-		if ascendErr != nil {
+			metadata.TableMetadata = append(metadata.TableMetadata, tableMeta)
 			return ascendErr
+		}(); err != nil {
+			return err
 		}
-		metadata.TableMetadata = append(metadata.TableMetadata, tableMeta)
 	}
 	footer, err := metadata.MarshalVT()
 	if err != nil {

